ls1intum/Artemis
# `Assessment`: Add interval based grade key editing

### Checklist
#### General


- [x] I tested locally ~and was confirmed by another developer on a test server~.
#### Client

### Motivation and Context


The current grading key has lots of input fields that makes it hard for the instructors to fill. Furthermore, seeing points and percentages together in the same row can be tiresome. This PR introduces a modified version of the existing grade key editing UI that defines the grade steps by point or percentage intervals. This reduces the number of inputs that the instructor needs to fill for each grade step by one and simplifies the screen in general by only showing points or percentages (not both) according to the current selection.
Additionally, this PR displays the upper bound percentage/point value of the highest grade step as infinity to avoid confusion. Because the existing implementation already maps grade values that are greater than the upper bound of the highest grade step to that step anyway so it is open ended, this PR makes it more explicit.

### Description

Interval based grade key editing mode is implemented such that the instructor only needs to fill the interval value and grade name for each grade step, lower and upper bounds are calculated automatically.

Sticky grade step at the bottom is added, it will be visible with the first added grade step. It is also made compatible with the generated default grade key.

Lower and upper bound percentage/point values for each grade step is now displayed with open/closed interval representation  e.g. [0-50) same as what will be shown in grade overview component. So the users can see the grade step bounds while editing instead of saving first and checking the grade key of a Course/Exam from its Statistics page.

`ModePickerComponent` was generalized to handle arbitrary modes as options. Previously the `ExerciseMode` enum was hardcoded as the only option.

`GradeStepBoundsPipe` is extracted from `grading-key-overview.component.html` to encapsulate the open/closed interval representation and make it reusable in other components.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis
2. Navigate to Course Administration
3. Select a Course or an Exam
4. Add a new Grading Key, modify it and save with different percentage and point combinations.
5. Ensure that interval based grading key is working consistent and reasonable.
6. Ensure that detailed grading key is working as the old grading key.
7. As a Student, go to the Statistics page of the Course or Exam mentioned above and then click View Grading Key / View Bonus Rules and check the grading key is consistent with what you saved in the steps above.
8.  Check the Mode Picker for changing the team / individual mode when creating exercises works as it was previously (due to refactoring).

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| base-grading-system.component.ts | 77.95% | 95.27% |
| detailed-grading-system.component.ts | 100% | 100% |
| grade-step-bounds.pipe.ts | 92.3% | 100% |
| grading-key-overview.component.ts | 74.46% | 84% |
| interval-grading-system.component.ts | 86.66% | 100% |
| team-config-form-group.component.ts | 100% | 100% |


### Screenshots


![image](https://user-images.githubusercontent.com/18427209/162097591-4cbc3896-3ede-476a-8730-d739fa2a0f96.png)
![image](https://user-images.githubusercontent.com/18427209/162097633-c5a14407-e163-4634-a56f-ec7ffa162da5.png)
![image](https://user-images.githubusercontent.com/18427209/162097944-ad72b183-d357-4453-a8d2-7fb69abe3b6b.png)
